### PR TITLE
Storage Providers 

### PR DESCRIPTION
**Pull request by jmoffatt32**
# Implementation of Storage Providers
DocPrompt currently supports loading a PDF document from a local file path. However, there is no way to save a document node, along with its associated metadata, and then reload the node in future executions. This can be solved in the short term by pickling nodes, however, if any modifications are made the the custom `DocumentNodeMetadata` model, then this pickle file becomes invalid. 

This PR seeks to create a cleaner implementation of persistant storage, through implementing a `storage` module and an `AbstractStorageProvider` interface. This abstract interface exposes a `store` and `persist` method, which let you seamlessly store and reload document nodes across multiple executions, without having to manually reassemble the metadata.

### Task List
- [x] Define base `fsspec` wrapper for reading/writing/deleting files using `fsspec`
- [x] Implement `from_storage` class method on `DocumentNode` for loading a document node from storage
- [x] Implement `persist` instance method on `DocumentNode` for saving a document node to storage
- [x] Define a custom base class for metadata, that should replace the current `Union[dict, BaseModel]` bound for the `PageNodeMetadata` and `DocumentNodeMetadata` generic type vars

### CODE REVIEW

File: docprompt/storage/_base.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    document_node_class: Annotated[
+        Type[DocumentNode], AfterValidator(validate_document_node_class)
+    ] = Field(..., repr=False)
+    document_metadata_class: Annotated[
+        Type[DocumentNodeMeatadata], AfterValidator(validate_document_metadata_class)
+    ] = Field(..., repr=False)
+
+    @abstractmethod
+    def store(self, document_node: DocumentNode) -> None:
```
Comment (PSU3D0) [inline docprompt/storage/_base.py:69]: This should support `**kwargs` for future extensibility

File: docprompt/storage/_base.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        """Store the document node in the storage provider.
+
+        Args:
+            document_node: The document node to store
+        """
+
+    @abstractmethod
+    def retrieve(self, file_hash: str) -> DocumentNode:
```
Comment (PSU3D0) [inline docprompt/storage/_base.py:77]: This should support `**kwargs` for future extensibility

File: docprompt/storage/_base.py (hunk 1 above)
Comment (PSU3D0) [reply docprompt/storage/_base.py:69]: Also some storages might return a resolved path return type should maybe be `Optional[Any]` or maybe even a generic

File: docprompt/storage/local.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+    Attributes:
+        base_file_path: The base file path for the document node
+        pdf: The file path for the pdf bytes
+        metadata: The file path for the metadata
+    """
+
+    base_file_path: Annotated[str, AfterValidator(validate_file_path)] = Field(...)
+
+    @property
```
Comment (PSU3D0) [inline docprompt/storage/local.py:47]: Should these be `computed_fields`?

File: docprompt/schema/storage.py
Comment (PSU3D0) [inline docprompt/schema/storage.py:1]: Extra file?

**Review by PSU3D0 — CHANGES_REQUESTED** [review]

File: docprompt/storage/_base.py (hunk 1 above)
Comment (jmoffatt32) [reply docprompt/storage/_base.py:69]: I like idea for generic here, will update the LocalProvider to return the sidecar. S3 could return a similar model with the bucket URIs

File: docprompt/storage.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        data["path"] = file_system[1]
+        data["fs_kwargs"] = fs_kwargs
+
+        return data
+
+    @model_validator(mode="after")
+    def validate_base_path_exists(self) -> Self:
+        """Validate that the base path exists after instantiating the file system."""
+        if not self.fs.exists(self.path, **self.fs_kwargs):
+            self.fs.mkdirs(self.path, **self.fs_kwargs)
```
Comment (PSU3D0) [inline docprompt/storage.py:118]: This means a filesystem is modified as a side effect of instantiation. Should this only be an explicit effect of write operations?

File: docprompt/schema/pipeline.py [hunk 5]
```diff
@@ -362,5 +544,5 @@ class DocumentCollection(
     Represents a collection of documents with some common metadata
     """
 
-    document_nodes: List[DocumentNode[DocumentNodeMetadata, PageNodeMetadata]]
-    metadata: DocumentCollectionMetadata
+    document_nodes: list[DocumentNode[DocumentNodeMetadata, PageNodeMetadata]]
```
Comment (PSU3D0) [inline docprompt/schema/pipeline.py:547]: To support python 3.8 needs to use List

**Review by PSU3D0 — CHANGES_REQUESTED** [review]
Looks good overall, few small things

File: docprompt/schema/metadata.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        else:
+            # We want to avoid setting any private attributes in the extra
+            # dictionary
+            if name.startswith("_"):
+                return super().__setattr__(name, value)
+
+            self.extra[name] = value
+
+
+class Metadata(BaseMetadata):
```
[the remark this answers is not in the document]
Comment (jmoffatt32) [reply docprompt/schema/metadata.py:136]: Good catch, this was a basic test class for script to check concrete implementations were working correct, will remove this

File: docprompt/schema/pipeline.py [hunk 7]
```diff
@@ -1,33 +1,43 @@
...
+DocumentCollectionMetadata = TypeVar("DocumentCollectionMetadata", bound=BaseMetadata)
+DocumentNodeMetadata = TypeVar("DocumentNodeMetadata", bound=BaseMetadata)
+PageNodeMetadata = TypeVar("PageNodeMetadata", bound=BaseMetadata)
 
-DocumentCollectionMetadata = TypeVar("DocumentCollectionMetadata", bound=BaseModel)
-DocumentNodeMetadata = TypeVar("DocumentNodeMetadata", bound=BaseModel)
-PageNodeMetadata = TypeVar("PageNodeMetadata", bound=BaseModel)
+# TODO: Is there a better way to bound these types?
+FilePathResult = TypeVar("FilePathResult", bound=BaseModel)
+StorageProvider = TypeVar("StorageProvider", bound=BaseModel)
```
[the remark this answers is not in the document]
Comment (jmoffatt32) [reply docprompt/schema/pipeline.py:40]: Good note, will also bound them to the proper models for type checking

File: docprompt/storage.py (hunk 4 above)
Comment (jmoffatt32) [reply docprompt/storage.py:118]: Yes, with new implementation of private `_wrtie` method, we can probably drop this post validator. Will remove and update unit tests as well to reflect changes

**Review by jmoffatt32 — COMMENTED** [review]
Cleaned up based on notes, and updated tests accordingly

File: docprompt/schema/metadata.py [hunk 8]
```diff
@@ -118,23 +131,19 @@ def __len__(self):
     def __getattr__(self, name):
         if self._is_field_typed():
             return super().__getattr__(name)
-        else:
-            return self.extra.get(name)
+
+        if name.startswith("_"):
+            return super().__getattr__(name)
+
+        return self.extra.get(name)
```
Comment (PSU3D0) [inline docprompt/schema/metadata.py:138]: This should catch a KeyError and reraise an attribute error to reproduce the same functionality as attribute lookup

**Review by PSU3D0 — CHANGES_REQUESTED** [review]

File: docprompt/schema/metadata.py (hunk 8 above)
Comment (jmoffatt32) [reply docprompt/schema/metadata.py:138]: I like that, much cleaner

**Review by PSU3D0 — APPROVED** [review]

**Pull request merged**